### PR TITLE
Prevent infinite recurse when 2 groups are members of each other

### PR DESCRIPTION
**Pull request by c-silva**
Tries to solve issue already reported:
https://github.com/theforeman/ldap_fluff/issues/51

### CODE REVIEW

File: lib/ldap_fluff/ad_member_service.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ def _groups_from_ldap_data(payload)
     data = []
     if !payload.nil?
       first_level  = payload[:memberof]
-      total_groups = _walk_group_ancestry(first_level)
+      total_groups, b = _walk_group_ancestry(first_level, first_level)
```
Comment (domcleal) [inline lib/ldap_fluff/ad_member_service.rb]: a) Replace `b` with `_` if you're discarding the variable.
b) The equals signs aren't lined up any more.
c) The extra argument to `_walk_group_ancestry` isn't declared in its method.

File: lib/ldap_fluff/ad_member_service.rb [hunk 2]
```diff
@@ -32,12 +32,15 @@ def _walk_group_ancestry(group_dns = [])
     group_dns.each do |group_dn|
       search = @ldap.search(:base => group_dn, :scope => Net::LDAP::SearchScope_BaseObject, :attributes => ['memberof'])
       if !search.nil? && !search.first.nil?
-        group = search.first
-        set  += _walk_group_ancestry(group[:memberof])
-        set  += group[:memberof]
+        groups = search.first[:memberof] - known_groups
```
Comment (domcleal) [inline lib/ldap_fluff/ad_member_service.rb]: Could this just use the existing list of gruops?  The arrays look like they'll probably contain the same data - they're adding both `groups` and `next_level` on every iteration.

**Comment by domcleal** [comment]
Please also check the test failures, they appear at the moment to be because the method doesn't have a second argument defined.

File: lib/ldap_fluff/ad_member_service.rb (hunk 2 above)
Comment (c-silva) [inline lib/ldap_fluff/ad_member_service.rb]: I'm not sure of what you mean by the existing list of groups. If you mean the `set` variable, it is empty when you recursevly re-enter`_walk_group_ancestry`.
Only by the end of all recursions will the arrays contain the same data.

**Comment by c-silva** [comment]
Hi, yes, that was the reason. The tests are passed now.

**Comment by domcleal** [comment]
Great, thanks. Would you be willing to write a test for this, simulating the looped membership? If not I can add it.

File: lib/ldap_fluff/ad_member_service.rb [hunk 3]
```diff
@@ -19,25 +19,28 @@ def find_user_groups(uid)
...
+        groups                       = search.first[:memberof] - known_groups
+        known_groups                += groups
+        next_level, new_known_groups = _walk_group_ancestry(groups, known_groups)
+        set                         += next_level
+        set                         += groups
+        known_groups                += next_level
       end
     end
-    set
+    return [set, known_groups]
```
Comment (dLobatog) [inline lib/ldap_fluff/ad_member_service.rb]: nitpick, `return` is optional here

Comment (c-silva) [inline lib/ldap_fluff/ad_member_service.rb]: ok!

**Comment by c-silva** [comment]
Hi, take a look. Hope it fits the purpose.

**Comment by domcleal** [comment]
Perfect, thanks @c-silva!

**Pull request merged**